plugin/transfer/transfer.go

```diff
@@ -0,0 +1,144 @@
+package transfer
+
+import (
+	"context"
+	"sync"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+var log = clog.NewWithPlugin(pluginName)
+
+// Transfer is a plugin that handles zone transfers.
+type Transfer struct {
+	xfrs []*xfr
+	Next plugin.Handler
+}
+
+type xfr struct {
+	Zones       []string
+	to          []string
+	Transferers map[string]Transferer // a map of zone->plugins which implement Transferer
+}
+
+// Transferer may be implemented by plugins to enable zone transfers
+type Transferer interface {
+	// AllRecords must return the SOA and all other records in the zone, including NS records
+	// for the given zone.
+	AllRecords(zone string) (*dns.SOA, []dns.RR, error)
```

Comment:
This is, I think, not the correct interface. It needs to be more like an iterator.

```
type Transferer interface {
   Walk(zone string, serial uint32) ([]dns.RR, error)
}
~~~
Or something like that, see latest tweaks in `plugin/file/xfr.go`
Two things that:
1. First RR returned must be a SOA
2. `io.EOF` signals the end of the list

Reply:
What about returning a channel?

Reply:
Ah, a channel, so we can start responding to the client in parallel with the plugin generating the records.  Yeah, makes sense.